designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,226 @@
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    processor: "markdown/markdown"
+};
+```
+
+This configuration file will apply the named processor `markdown` from the `eslint-plugin-markdown` plugin to all files.
+
+The `processor` option would be interpreted inside of `CLIEngine` to determine which processors to apply to files before linting should begin.
+
+### Apply Processors Using `overrides`
+
+Of course, in most cases the end-user will not want to apply a processor to all files. In that case, the `overrides` configuration option may be used to narrowly define to which files the processors should be applied. For example:
+
+```js
+module.exports = {
+    plugins: ["markdown"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    overrides: [
+        {
+            files: ["*.md"],
+            processor: "markdown/markdown"
+        }
+    ]
+};
+```
+
+In this configuration, the named processor `markdown` from the `eslint-plugin-markdown` plugin is applied only to files with the extension `*.md`. This mimics the behavior of the extension-named processor that `eslint-plugin-markdown` currently exports.
+
+The `CLIEngine` method `getConfigForFile()` would automatically get the correct value for `processor` when the configuration is calculated.
+
+**Note:** End-users would still need to provide `--ext .md` on the command line for ESLint to automatically lint files ending with `.md`. No part of this proposal changes the behavior of `--ext`.
+
+### Virtual Filenames for Code Blocks
+
+The `preprocess()` method inside of a processor can now return an array containing either strings (the current behavior) or an object with two properties:
+
+* `filePath` - (string) a virtual file path to map the code block to
+* `text` - (string) the source code text
+
+For example:
+
+```js
+module.exports = {
+    preprocess(text, filename) {
+        
+        const { jsText1, jsText2 } = doSomething(text, filename);
+        
+        return [
+            {
+                filePath: filename + "1.js",
+                text: jsText1
+            },
+            {
+                filePath: filename + "2.js",
+                text: jsText2
+            }
+        ]
+    }
+};
+```
+
+This processor returns two code blocks containing JavaScript code. Each code block is given a virtual filename ending with `.js`.
+
+When a `preprocess()` method returns an object with a `filePath` property, `CLIEngine` will call `getConfigForFile()` on the `filePath` property to determine the correct configuration for the code block, which includes whether another processor should be run on the code block (matched by file extension).
```

Comment:
At least in the case of the Markdown processor, since it can’t know which extensions have processors configured, I understand it will want to return every code block it encounters? For example, if `original.md` contains `js`, `html`, and `vue` code blocks, it would return `original.md.js`, `original.md.html`, and `original.md.vue`. Do we then need to run those through the `--ext` filter as well? Perhaps I invoke with `--ext .html`, so I’d expect the `html` code block to be linted, but not the `vue` code block.

Reply:
At initial, I had considered the same thing: we should filter extracted blocks with the `--ext` or glob patterns. However, I think currently, if a specific file specified (E.g., `eslint docs/README.md`), I'm not sure what the proper behavior is. I will be confused if ESLint changes the target code blocks by globs. So I want options for processors in config files.

Reply:
I think that makes the most sense. The Markdown processor would return every fenced code block, and the HTML processor would return every `<script>` tag, and ESLint core would be responsible for deciding which of those get linted.

How that interacts with `--ext` I’m not sure. I suppose you could argue that if I configure an `override` for `*.ts`, then I shouldn’t also have to specify `--ext .ts`. When plugins provided their own extensions, the `--ext` flag was necessary, but flipping the relationship around so that the extensions are configured in `overrides` might make it redundant.

Is appending an extension to the existing filename the best way? I think so, but I wanted to point out that it’s trivial for the Markdown processor to pull the tag from the fenced code block, but an HTML processor would need to map from the `<script>`’s MIME type to the appropriate extension.

Reply:
I don't oppose the way.

My small concern is, some users use glob patterns rather than `--ext` option to specify other files than JS (E.g., `src/**/*.{js,ts,tsx}`). But we can say that use `--ext` option to specify embedded codes other than JS.

Reply:
Okay, it seems like we are in agreement that:

1. Processors can return all file types it thinks ESLint can handle.
2. ESLint will filter out any file types that aren't specified by `--ext` to avoid errors.
3. Users must use `--ext` even if they have `overrides` in their config; `overrides` is not a replacement for `--ext` (in the context of this proposal).

@btmills I think appending the file extension makes sense. My hunch is that most processors will only be returning `.js` files any way. As @mysticatea has pointed out a couple of times, we are really talking about an edge case with non-`.js` code blocks being returned from processors.